Full style rebuild when window moves to foreground or background

mwyrzykowski opened the pull request:
#### 52124d1458b16c0ad3f88d7aeef18a0ca3c3ec2b
<pre>
Full style rebuild when window moves to foreground or background
<a href="https://bugs.webkit.org/show_bug.cgi?id=297610">https://bugs.webkit.org/show_bug.cgi?id=297610</a>
<a href="https://rdar.apple.com/158668873">rdar://158668873</a>

Reviewed by Antti Koivisto.

Skip style recalculation when none of the members of ScreenProperties
which impact style change.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::screenPropertiesDidChange):
* Source/WebCore/page/Page.h:
* Source/WebCore/platform/PlatformScreen.h:
* Source/WebCore/platform/ScreenProperties.h:
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::setScreenProperties):

Canonical link: <a href="https://commits.webkit.org/299041@main">https://commits.webkit.org/299041@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e096b94999109478e0db1ba2e0b8f0c1af9ad904

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117531 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37207 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/27831 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/123649 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/69538 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/34fcddc6-b79d-4474-9eb5-064ce06aafcc) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/119409 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/37901 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/45791 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89205 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-transforms/transform-box/cssbox-content-box-002.html (failure)") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/45019 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/b7d40a38-99b4-464e-a41d-454ebf88a610) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120483 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30182 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105388 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/69711 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/31e4e899-19d7-479d-a176-afdd4b04939b) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29243 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/23507 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67308 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99597 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23688 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126756 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44434 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33433 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97872 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/44792 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101619 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97660 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24867 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43031 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20975 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/40792 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44305 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/49980 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/43762 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/47111 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45455 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/7367fbdf06b79dc6aa226e27fa7451b7611fb8b1. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49606

squelart commented on Source/WebCore/html/canvas/GPUCanvasContextCocoa.mm:
```diff
@@ -210,12 +210,6 @@ static float interpolateHeadroom(float headroomForLow, float headroomForHigh, fl
 
     auto limitValue = m_dynamicRangeLimit.value();
 
-    if (m_suppressEDR) {
```
Another way to think about this: `dynamic-range-limit` and HDR-suppression are independent/parallel ways to constrain HDR down from the physical screen headroom, and we should use the minimum of those two to compute the effective headroom limit.
So we can't just apply HDR-suppression silently, and then hope to compute the `dynamic-range-limit` contribution from that "current headroom" that may or may not be suppressed already.

anttijk commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1385,7 +1387,8 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
         return;
 
     m_suppressEDR = suppressEDR;
-    screenPropertiesChanged();
+    constexpr bool skipStyleRecalc = true;
+    screenPropertiesChanged(skipStyleRecalc);
```
If `suppressEDR` bit is not part of the ScreenProperties then screenPropertiesChanged message would only be needed if the computed `currentEDRHeadroom` actually changes here. But at least in my system in MiniBrowser `currentEDRHeadroom` always computes to "1" so we shouldn't be sending any messages here.

mwyrzykowski replied:
Is your display set to XDR color profile? That might explain the difference

anttijk replied:
That's not really the point. If nothing changes then we should not be sending ScreenProperties change messages.

anttijk commented on Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
```diff
@@ -2494,7 +2494,6 @@ header: <WebCore/ScreenProperties.h>
     bool screenSupportsHighDynamicRange;
 #if HAVE(SUPPORT_HDR_DISPLAY)
-    bool suppressEDR;
     float currentEDRHeadroom;
     float maxEDRHeadroom;
```
Why do these things that depend on application foreground status need to be in a `ScreenProperties` isn the first place? That's what is causing these problems. Can't the suppression status/current headroom be passed separately as a separate message that would only do the required work and not all the work associated with screen property changes?

mwyrzykowski replied:
We could move currentEDRHeadroom out of ScreenProperties but would likely want to discuss with @mattwoodrow first

mwyrzykowski commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1264,7 +1264,7 @@ void setLockdownModeEnabledGloballyForTesting(std::optional<bool> enabledForTest
 #endif // ENABLE(CFPREFS_DIRECT_MODE)
 
-void WebProcessPool::screenPropertiesChanged()
+void WebProcessPool::screenPropertiesChanged(bool skipStyleRecalc)
```
Oh that is a good idea

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/446b7b23f6dd743a2719285c2af662ff39275e33. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49606

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/61214ddee2cbeef238da6a1ad0c8aacdd28d8f7a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49606

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e096b94999109478e0db1ba2e0b8f0c1af9ad904. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49606

mwyrzykowski committed:
Full style rebuild when window moves to foreground or background
https://bugs.webkit.org/show_bug.cgi?id=297610
rdar://158668873

Reviewed by Antti Koivisto.

Skip style recalculation when none of the members of ScreenProperties
which impact style change.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::screenPropertiesDidChange):
* Source/WebCore/page/Page.h:
* Source/WebCore/platform/PlatformScreen.h:
* Source/WebCore/platform/ScreenProperties.h:
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::setScreenProperties):

Canonical link: https://commits.webkit.org/299041@main
